thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/DefaultTimeSeriesCache.java

```diff
@@ -0,0 +1,154 @@
+  private void fetchMissingSlices(ThirdEyeCacheResponse cacheResponse) throws Exception {
+
+    ThirdEyeRequest request = cacheResponse.getCacheRequest().getRequest();
+    long metricId = request.getMetricFunctions().get(0).getMetricId();
+    long requestSliceStart = request.getStartTimeInclusive().getMillis();
+    long requestSliceEnd = request.getEndTimeExclusive().getMillis();
+
+    ThirdEyeResponse result;
+    MetricSlice slice;
+
+    if (cacheResponse.isMissingStartSlice(requestSliceStart)) {
+      slice = MetricSlice.from(metricId, requestSliceStart, cacheResponse.getFirstTimestamp(), request.getFilterSet(), request.getGroupByTimeGranularity());
+      result = fetchSliceFromSource(slice);
+      insertTimeSeriesIntoCache(result);
+      cacheResponse.mergeSliceIntoRows(result, MergeSliceType.PREPEND);
+    }
+
+    if (cacheResponse.isMissingEndSlice(requestSliceEnd)) {
+      // we add one time granularity to start because the start is inclusive.
+      slice = MetricSlice.from(metricId, cacheResponse.getLastTimestamp() + request.getGroupByTimeGranularity().toMillis(), requestSliceEnd, request.getFilterSet(), request.getGroupByTimeGranularity());
+      result = fetchSliceFromSource(slice);
+      insertTimeSeriesIntoCache(result);
+      cacheResponse.mergeSliceIntoRows(result, MergeSliceType.APPEND);
+    }
+  }
+
+  private ThirdEyeResponse fetchSliceFromSource(MetricSlice slice) throws Exception {
+    TimeSeriesRequestContainer rc = DataFrameUtils.makeTimeSeriesRequestAligned(slice, "ref", this.metricDAO, this.datasetDAO);
+    return this.queryCache.getQueryResult(rc.getRequest());
+  }
+
+  private ThirdEyeResponse buildThirdEyeResponseFromCacheResponse(ThirdEyeCacheResponse cacheResponse) {
+
+    List<String[]> rows = new ArrayList<>();
+    ThirdEyeRequest request = cacheResponse.getCacheRequest().getRequest();
+
+    String dataset = request.getMetricFunctions().get(0).getDataset();
+    DatasetConfigDTO datasetDTO = datasetDAO.findByDataset(dataset);
+    TimeSpec timeSpec = ThirdEyeUtils.getTimeSpecFromDatasetConfig(datasetDTO);
+    DateTimeZone timeZone = DateTimeZone.forID(datasetDTO.getTimezone());
+
+    for (TimeSeriesDataPoint dataPoint : cacheResponse.getRows()) {
+      int timeBucketIndex = TimeRangeUtils.computeBucketIndex(
+          request.getGroupByTimeGranularity(), request.getStartTimeInclusive(), new DateTime(dataPoint.getTimestamp(), timeZone));
+      String dataValue = dataPoint.getDataValue();
+
+      String[] row = new String[2];
+      row[0] = String.valueOf(timeBucketIndex);
+      row[1] = (dataValue == null || dataValue.equals("null")) ? "0" : dataValue;
+
+      rows.add(row);
+    }
+
+    return new RelationalThirdEyeResponse(request, rows, timeSpec);
+  }
+
+  public void insertTimeSeriesIntoCache(ThirdEyeResponse response) {
+
+    // we can consider using fixedThreadPool instead.
+    ExecutorService executor = Executors.newCachedThreadPool();
+
+    // insert points in parallel
```

Comment:
Any reason we want to insert points in parallel? Is the inserting very slow?

Reply:
It's not overly slow, but we still save a couple seconds by parallelizing this. since each data point has no dependency on other data points (as in order of insertion doesn't matter), I figured this would be an easy optimization.